photon-api/src/main/scala/com/linkedin/photon/ml/data/LocalDataset.scala

```diff
@@ -188,61 +188,69 @@ object LocalDataset {
   protected[ml] def computePearsonCorrelationScore(
       labelAndFeatures: Array[(Double, Vector[Double])]): Map[Int, Double] = {
 
-    val featureLabelProductSums = mutable.Map[Int, Double]()
-    val featureFirstOrderSums = mutable.Map[Int, Double]()
-    val featureSecondOrderSums = mutable.Map[Int, Double]()
-    var labelFirstOrderSum = 0.0
-    var labelSecondOrderSum = 0.0
-    var numSamples = 0
+    val featureMeans = mutable.Map[Int, Double]()
+    var featureUnscaledVars = mutable.Map[Int, Double]()
+    var labelMean = 0.0
+    var labelUnscaledVar = 0.0
+    val unscaledCovariances = mutable.Map[Int, Double]()
     var interceptAdded = false
+    var numSamples = 0
 
     labelAndFeatures.foreach { case (label, features) =>
       numSamples += 1
-      labelFirstOrderSum += label
-      labelSecondOrderSum += label * label
+
+      val deltaLabel = label - labelMean
+      labelMean += deltaLabel / numSamples
+      labelUnscaledVar += deltaLabel * (label - labelMean)
+
       // Note that, if there is duplicated keys in the feature vector, then the following Pearson correlation scores
       // calculation will screw up
       features.activeIterator.foreach { case (key, value) =>
```

Comment:
My major concern is on this `activeIterator` here. In the old implementation, we could simple ignore the zeros for sparse features. However, the new implementation should take zeros into consideration. Then first, I will dramatically increase the computational cost by treating a large sparse vector as dense. Also, this implementation might be wrong because the `activeIterator` seems to take only non-zero values.

Reply:
This can be easily verified with some sparse features in the Unit Test.

Reply:
Nice catch, and thanks for the comment!